### PR TITLE
fix(#971): parse submodule .git files instead of erroring

### PR DESCRIPTION
**Pull request by Kintar**
Altered logic for detecting git directory. Instead of erroring on non-direcctory .git files, now parses the file and returns the linked gitdir.

Note that since the module is a different repository, you have to `git bug user new` in the submodule before using git-bug. I've never used working trees, so I haven't tested that.

Also, I'm brand new to this project and haven't quite worked out how to add a unit test for this. Happy to do that if needed before this PR can be accepted.

### CODE REVIEW

File: repository/gogit.go [hunk 1]
```diff
@@ -170,6 +171,28 @@ func detectGitPath(path string) (string, error) {
...
+				// clean up before returning an error
+				reader := bufio.NewReader(dotfile)
+				line, _, err := reader.ReadLine()
+				if err != nil {
+					_ = dotfile.Close()
+					return "", fmt.Errorf(".git exists but is not a direcctory and cannot be read: %w", err)
+				}
+				dotContent := string(line)
+				if strings.HasPrefix(dotContent, "gitdir:") {
+					_ = dotfile.Close()
```
Comment (MichaelMure) [inline repository/gogit.go:190]: This `Close()` is misplaced, you want to close as soon as possible (just after the `ReadLine()` here, after handling the error). In your code still have one code path that can lead to not closing the file.

File: repository/gogit.go [hunk 2]
```diff
@@ -170,6 +171,28 @@ func detectGitPath(path string) (string, error) {
...
+					// This is a submodule parent path link. Strip the prefix, clean the string of whitespace just to
+					// be safe, and return
+					dotContent = strings.TrimSpace(dotContent[7:])
```
Comment (MichaelMure) [inline repository/gogit.go:193]: It's not wrong, but what about using `strings.TrimPrefix(dotContent, "gitdir:")` that tells a bit more what you are doing?

**Review by MichaelMure — CHANGES_REQUESTED** [review]
Small nitpicking.
It'd be nice indeed to have some test for that. Maybe add a new test case in https://github.com/MichaelMure/git-bug/blob/60d40d60e9f91247b61f541888f1469bff44f573/repository/gogit_test.go#L12 ? That is:
- generate a repo as normal
- create a temp folder (`t.TempDir()`), put the proper .git file there
- make sure that opens properly

**Comment by MichaelMure** [comment]
> Note that since the module is a different repository, you have to `git bug user new` in the submodule before using git-bug. I've never used working trees, so I haven't tested that.

Could you expand a bit on that? Once your code setup the root at the "remote" repository, there should be no difference with a normal repo.

File: repository/gogit.go (hunk 2 above)
Comment (Kintar) [reply repository/gogit.go:193]: Would you believe I've been using Go since 1.8 and never noticed the strings.TrimPrefix function? XD  Absolutely!

**Comment by Kintar** [comment]
@MichaelMure said:

> Could you expand a bit on that?

Sure.  Say you have a repository "repo1" with a submodule "module1".  If you're in a directory which is tracked by `repo1` then the git root directory is `repo1/.git/`, and git-bug's metadata lives in `repo1/.git/git-bug/`.  If you then switch to a directory tracked by the submodule then the git root becomes `repo1/.git/modules/module1`, and git-bug will start using metadata from `repo1/.git/modules/module1/git-bug`.

I think this is the desired behavior, since technically you're now tracking bugs on the submodule and not on the parent repository, but I called it out in case the assumption was made that bugs logged while in a submodule path should be tracked in the parent repository.  If that's the case, then the logic in gogit.detectGitPath needs to return the parent repository's git path and not the submodule's.

**Comment by Kintar** [comment]
Changes submitted as requested.

**Comment by Kintar** [comment]
Realized I did not update the error message returned when .git is neither a directory nor a workspace file. Updated error message.

**Comment by MichaelMure** [comment]
Sorry for the delay.

I improved your PR a bit:
- add a limited reader to avoid attacks with very large .git file
- support recursive indirection (up to depth 10)
- check that the pointed to repo does exist

**Pull request merged**